index.d.ts

```diff
@@ -0,0 +1,60 @@
+export interface Action {
+  type: string;
+}
+
+export type Reducer<S> = (state: S, action: Action) => S;
+
+export type Dispatch = (action: any) => any;
```

Comment:
How do we use it? Considering first option we'd have:
```ts
export interface Store<S> {
  dispatch: Dispatch<any>;
  // ...
}
```

So the only place where type parameters would be useful is the app code, e.g.
```ts
@connect(
  state => /* ... */,
  (dispatch: Dispatch<Action>) => {
    return {
      someActionCreator: () => dispatch(someActionCreator)
    }
  }
)
```

But nothing prevents you to set up custom `Dispatch` type with constraints that fit you needs:
```ts
type MyDispatch = (action: MyAction) => MyAction;
```
Then you can use your strict `MyDispatch` in place of `Dispatch` anywhere and it will type-check correctly.

IMO if we add type parameters here, we'd have to write `Dispatch<any>` most of the time, because concrete type constraints depend on the middleware(s) you use.